src/Storages/MergeTree/ReplicatedMergeTreeTableMetadata.cpp

```diff
@@ -36,6 +39,21 @@ static String formattedASTNormalized(const ASTPtr & ast)
     return buf.str();
 }
 
+static bool peekString(const char * s, ReadBufferFromString & buf)
+{
+    auto * current_pos = buf.position();
+    for (; *s; ++s)
+    {
+        if (buf.eof() || *buf.position() != *s)
+        {
+            buf.position() = current_pos;
```

Comment:
It's unsafe to use a previously saved position after checking for eof, because `eof()` might call `nextImpl()` which might refill the buffer with a new bunch of data or even reallocate the buffer. See `PeekableReadBuffer`

Reply:
> because eof() might call nextImpl() which might refill the buffer with a new bunch of data or even reallocate the buffer
Yes, I'm aware of this so I only implement `peekString` for `ReadBufferFromString`. But now look at `ReadBufferFromString` code seem it doesn't work as I thought, still have internal buffer. Then this implementation is incorrect.